File: src/com/timepath/EnumFlags.java
package com.timepath;

import java.util.Arrays;
import java.util.Comparator;
import java.util.EnumSet;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * @author TimePath
 */
public final class EnumFlags {

    private static final Logger LOG = Logger.getLogger(EnumFlags.class.getName());

    private EnumFlags() {
    }

    @SuppressWarnings("unchecked")
    public static <C extends Enum<C> & EnumFlag> EnumSet<C> decode(int encoded, Class<C> enumClass) {
        C[] map = enumClass.getEnumConstants();
        // Mixed bits at the top, single bits at the bottom, in order of 1s
        Arrays.sort(map, new Comparator<C>() {
            @Override
            public int compare(C e1, C e2) {
                int i1 = e1.getId();
                int i2 = e2.getId();
                int diff = Integer.bitCount(i2) - Integer.bitCount(i1);
                return ( diff == 0 ) ? ( i2 - i1 ) : diff; // If same amount of bits, higher value first
            }
        });
        for(C entry : map) {
            if(entry.getId() == encoded) {
                LOG.log(Level.FINER, "{0} = {1}", new Object[] { entry, encoded });
                return EnumSet.of(entry);
            }
        }
        EnumSet<C> ret = EnumSet.noneOf(enumClass);
        for(C entry : map) {
            if(entry.getId() == 0) {
                continue;
            }
            if(( encoded & entry.getId() ) == entry.getId()) {
                ret.add(entry);
            }
        }
        LOG.log(Level.FINER, "{0} = {1}", new Object[] { ret, encoded });
        return ret;
    }

    public static <C extends Enum<C>> int encode(Iterable<C> set) {
        int ret = 0;
        for(C val : set) {
            ret |= 1 << val.ordinal();
        }
        return ret;
    }
}

File: src/main/kotlin/com/timepath/EnumFlags.kt
package com.timepath


import java.util.*

/** Implement the [EnumFlags.Flag] interface to use these utilities */
public object EnumFlags {

    public interface Flag<E : Enum<E>> {
        public val id: Int get() = 1 shl (this as Enum<*>).ordinal
    }

    public fun <E> decode(encoded: Int, enumClass: Class<E>): EnumSet<E> where E : Flag<E>, E : Enum<E> {
        val map = enumClass.enumConstants
        // Mixed bits at the top, single bits at the bottom, in order of 1s
        Arrays.sort<E>(map, Comparator { e1, e2 ->
            val i1 = e1.id
            val i2 = e2.id
            val diff = Integer.bitCount(i2) - Integer.bitCount(i1)
            when (diff) {
                0 -> i2 - i1 // If same amount of bits, higher value first
                else -> diff
            }
        })
        // Try an exact match
        for (entry in map) {
            if (entry.id == encoded) return EnumSet.of(entry)
        }
        // Aggregate
        val ret = EnumSet.noneOf(enumClass)
        for (entry in map) {
            if (entry.id == 0) continue
            if ((encoded and entry.id) == entry.id) ret.add(entry)
        }
        return ret
    }

    public fun <E : Flag<E>> encode(set: Iterable<E>): Int {
        var ret = 0
        for (flag in set) {
            ret = ret or flag.id
        }
        return ret
    }
}
